File: typescript/src/workshop/m4/s07_movemethod/step1/BookingPrinter.ts
import type { Booking } from './Booking.js';
import type { Screening } from './Screening.js';

/**
 * Krok 1: Move Method `screeningLine` -> `Screening.headline()`.
 * Metoda używała tylko danych seansu, więc właścicielem jest Screening.
 * Wywołanie po przeniesieniu czyta się `booking.screening.headline()`.
 */
export class BookingPrinter {
  print(booking: Booking): string {
    return 'Rezerwacja ' + booking.id + '\n'
      + booking.screening.headline() + '\n'
      + 'Miejsce: ' + booking.seat + '\n'
      + 'Pozostale wolne: [' + this.remainingSeats(booking.screening, booking.seat).join(', ') + ']\n';
  }

  private remainingSeats(s: Screening, seat: number): number[] {
    const free = [...s.freeSeats];
    free.splice(free.indexOf(seat), 1);
    return free;
  }
}

File: csharp/src/Training.Workshop/M4/S07MoveMethod/Step2/BookingPrinter.cs
namespace Training.Workshop.M4.S07MoveMethod.Step2;

/// <summary>
/// Krok 2 (rozwiązanie): Move Method <c>RemainingSeats</c>
/// -&gt; <c>Screening.FreeSeatsWithout(int?)</c>. BookingPrinter zostaje koordynatorem wydruku:
/// składa tekst z tego, co wiedzą Booking i Screening.
/// </summary>
public sealed class BookingPrinter
{
    public string Print(Booking booking)
    {
        return "Rezerwacja " + booking.Id + "\n"
            + booking.Screening.Headline() + "\n"
            + "Miejsce: " + booking.Seat + "\n"
            + "Pozostale wolne: " + booking.Screening.FreeSeatsWithout(booking.Seat) + "\n";
    }
}
